### PR TITLE
CELDEV-787 JS EventManager

### PR DESCRIPTION
**Pull request by msladek**
https://synjira.atlassian.net/browse/CELDEV-787

### CODE REVIEW

File: src/main/webapp/resources/celJS/initCelements.js [hunk 1]
```diff
@@ -638,6 +644,214 @@
...
+        console.debug('EventHandler - unregister: ', _me._eventName, _me._cssSelector,
+          _me._className, _me._actionFunction.name, _me._htmlElement);
+      }
+  
+    });
+  }
+
+  if(typeof window.CELEMENTS.EventManager === 'undefined') {
+    window.CELEMENTS.EventManager = Class.create({
+      _instructionRegex : new RegExp('([\\w]+)([%+-])([\\w-]+):([^?]+)\\??(.+)?'),
```
Comment (fpichler) [inline src/main/webapp/resources/celJS/initCelements.js]:
```suggestion
      _instructionRegex : new RegExp('([\\w:]+)([%+-])([\\w-]+):([^?]+)\\??(.+)?'),
```

File: src/main/webapp/resources/celJS/initCelements.js [hunk 2]
```diff
@@ -638,6 +644,214 @@
...
+        var _me = this;
+        var parts = instruction.match(_me._instructionRegex) || [];
+        var data = {
+          eventName : parts[1],
+          action : parts[2],
+          className : parts[3],
+          cssSelector : parts[4],
+          condition : parts[5]
+        };
+        if (data.eventName === 'load') {
```
Comment (fpichler) [inline src/main/webapp/resources/celJS/initCelements.js]: you can delete this block. The developer should directly use the cel:-event. Please add this functionality to the documentation.

File: src/main/webapp/resources/celJS/initCelements.js [hunk 3]
```diff
@@ -638,6 +644,214 @@
...
+            }
+          });
+          if (newElem.eventHandlers.length > 0) {
+            _me._eventElements.push(newElem);
+            console.debug(logPref, 'new element ', htmlElem);
+          } else {
+            console.debug(logPref, 'no valid instructions found on ', htmlElem);
+          }
+          htmlElem.addClassName('celOnEventInitialized');
+          htmlElem.fire('cel:onEventInitialized');
```
Comment (fpichler) [inline src/main/webapp/resources/celJS/initCelements.js]: please use shorter event-name for example:
```suggestion
          htmlElem.fire('celEM:init');
```

**Review by fpichler — CHANGES_REQUESTED** [review]

File: src/main/webapp/resources/celJS/initCelements.js [hunk 4]
```diff
@@ -831,10 +827,10 @@
           var elem = _me._eventElements[i];
           var isInBody = $(document.body).contains(elem.htmlElem);
           var changedDataValue = (elem.htmlElem.readAttribute('data-cel-event') !== elem.dataValue);
-          if (!isInBody || changedDataValue) {
+          if (!isInBody || changedDataValue || !elem.htmlElem.hasClassName('celOnEventInit')) {
```
Comment (fpichler) [inline src/main/webapp/resources/celJS/initCelements.js:830]: For which possible situation is the check on classname needed? The contract is, that elements with the celOnEventInit class are initalized and thus have a representation in the _me._eventElements the class is added after initialization and removed only here after resetting the handlers.

**Review by fpichler — CHANGES_REQUESTED** [review]

File: src/main/webapp/resources/celJS/initCelements.js (hunk 4 above)
Comment (msladek) [reply src/main/webapp/resources/celJS/initCelements.js:830]: the check ensures correct handling if the init class was removed. e.g. manually when developing new layouts to test out functionality.
without it the element could be contained multiple times in the elements list.
the additional check doesn't break the contract but makes the implementation more robust.

Comment (fpichler) [reply src/main/webapp/resources/celJS/initCelements.js:830]: 👍

**Review by fpichler — APPROVED** [review]

**Pull request merged**